dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/preheat/supplier/PreheatClassScanner.java

```diff
@@ -0,0 +1,109 @@
+ * Redistributions of source code must retain the above copyright notice, this
+ * list of conditions and the following disclaimer.
+ *
+ * Redistributions in binary form must reproduce the above copyright notice,
+ * this list of conditions and the following disclaimer in the documentation
+ * and/or other materials provided with the distribution.
+ * Neither the name of the HISP project nor the names of its contributors may
+ * be used to endorse or promote products derived from this software without
+ * specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+ * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR
+ * ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON
+ * ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+import org.hisp.dhis.tracker.preheat.supplier.classStrategy.StrategyFor;
+
+import io.github.classgraph.AnnotationClassRef;
+import io.github.classgraph.AnnotationInfo;
+import io.github.classgraph.ClassGraph;
+import io.github.classgraph.ClassInfo;
+import io.github.classgraph.ScanResult;
+
+/**
+ * This class is responsible for creating an associative map where the key is
+ * the name of a Strategy class and the key is the name of the domain object
+ * class to cache (based on the {@see StrategyFor} annotation)
+ * 
+ * @author Luciano Fiandesio
+ */
+public class PreheatClassScanner
+{
+    public List<String> scanSuppliers()
+    {
+        final String pkg = getCurrentPackage();
+        final String annotation = SupplierDependsOn.class.getName();
+        try (ScanResult scanResult = new ClassGraph()
+            // .disableJarScanning()
+            .enableAllInfo()
+            .disableNestedJarScanning()
+            .acceptPackages( pkg )
+            .scan())
+        {
+            List<String> cls = scanResult.getAllClasses().stream()
+                .filter( c -> c.extendsSuperclass( AbstractPreheatSupplier.class.getName() ) )
+                .map( ClassInfo::getSimpleName ).collect( Collectors.toList() );
+
+            for ( ClassInfo classInfo : scanResult.getClassesWithAnnotation( annotation ) )
+            {
+                String dependsOn = getTargetClass( classInfo, annotation );
```

Comment:
Not sure if I'm just not thinking straight, but I'm not sure this for loop does what we want it to?

So, I understand it as if we want to sort the list so that dependencies comes in the correct order. IE. if A->B->C, we want to make sure the list follows this.

However, if we have an already ordered list, A->B->C, and C depends on A (which is still true in this order), we swap A and C, ending up out of order? So we need an if statement making sure we only swap when indexOf dependency is higher than current class?

Reply:
Removed entire class-path based logic